types.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2025 Pinterest
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package thriftcheck
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"go.uber.org/thriftrw/ast"
+)
+
+// TypeMatcher represents a way to match against AST nodes.
+type TypeMatcher interface {
+	Matches(c *C, n ast.Node) bool
+	Name() string
+}
+
+// ThriftType implements fig StringUnmarshaler for automatic toml parsing.
+type ThriftType struct {
+	name    string
+	matcher TypeMatcher
+}
+
+// UnmarshalString implements fig.StringUnmarshaler for automatic toml parsing.
+func (t *ThriftType) UnmarshalString(v string) error {
+	name := strings.ToLower(v)
+	factory, ok := typeFactories[name]
+	if !ok {
+		validTypes := make([]string, 0, len(typeFactories))
+		for k := range typeFactories {
+			validTypes = append(validTypes, k)
+		}
+		sort.Strings(validTypes)
+		return fmt.Errorf("unknown type: %s, valid types are: %v", v, validTypes)
+	}
+
+	t.name = name
+	t.matcher = factory(name)
+	return nil
+}
+
+func (t *ThriftType) Matches(c *C, n ast.Node) bool {
+	return t.matcher.Matches(c, n)
+}
+
+func (t *ThriftType) Name() string {
+	return t.name
+}
+
+// thriftTypeMatcher handles all types with unified TypeReference resolution.
+type thriftTypeMatcher struct {
+	name    string
```

Comment:
I don't think the matchers need to know their name anymore. `ThriftType` now has this information.